TC-CGEN-2.2 python test

juandediosg opened the pull request:
### Description
This PR is focused on implementing and validating the **TC-CGEN-2.2**
Fixes: https://github.com/project-chip/matter-test-scripts/issues/404
Test plan: https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/cluster/General_Commissioning.adoc

### Purpose ###
This test case validates the behavior of the failsafe timer during the commissioning process, focusing on its expiration, reset behavior, and interactions with commissioning-related commands. The test ensures that the failsafe behaves as expected in various scenarios and has been adapted to bypass waiting for the failsafe to expire, optimizing test execution in CI environments. This approach simulates real-world scenarios such as network misconfigurations, incorrect credentials, expired certificates, and other issues that might occur during commissioning.

### Notes ###
This PR includes a workaround for the waits related to the failsafe timer expiration. As specified in the test plan, the original behavior expected a wait for the failsafe expiry. However, to improve CI testing times and avoid unnecessary delays, the failsafe timer is set to `0` seconds using the function `expire_failsafe_timer()`.

### Testing ###
To run the automated TC-CGEN-2.2 test, follow these steps:

1. Ensure your environment is set up and the necessary dependencies are installed.
2. Run the following command from the command line:

`python3 src/python_testing/TC_CGEN_2_2.py --commissioning-method on-network --qr-code MT:-24J0AFN00KA0648G00`    

To simulate CI
`python3 src/python_testing/TC_CGEN_2_2.py --commissioning-method on-network --qr-code MT:-24J0AFN00KA0648G00 --PICS src/app/tests/suites/certification/ci-pics-values --int-arg PIXIT.CGEN.FailsafeExpiryLengthSeconds:1`

juandediosg committed:
Implement partial solution up to step 5

juandediosg committed:
Draft: Work in progress until step 22

juandediosg committed:
Draft PR - Verify assert values and finalize test case for TC_CGEN_2_2

juandediosg committed:
Improvements in failsafe handling, timeouts, and preparation for step 44 in TC_CGEN_2_2

juandediosg commented:
Currently exploring two improvements for the test:

- Handling the waits: The test has long waits, especially in the failsafe process. One option being explored is to minimize these waits in CI by arming the failsafe with a time=0 command, which would trigger the failsafe immediately without unnecessary delays.
- Step before 44: There’s also the need to add a step before 44 to remove TH2, which wasn’t accounted for in the original flow. This will ensure that the number of root certs returns to the expected state before proceeding.

juandediosg committed:
Implemented failsafe expiry workaround, cleanup for TC_CGEN_2_2, and step 44 updated

juandediosg committed:
Updated code-lints and fixed f-string warnings

juandediosg committed:
Fixed import order and to conflicts for Restyled

juandediosg committed:
Merge branch 'project-chip:master' into TC_CGEN_2_2_PythonTest

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,796 @@
+        # which immediately expires the failsafe timer (sets it to 0), bypassing the natural wait and speeding up the test execution.
+
+        # This function bypasses the wait for the FailsafeTimer to expire for TH2 as originally defined in the test plan.
+        resp = await self.expire_failsafe_timer(dev_ctrl=TH2, node_id=newNodeId+1)
```
So the point of step 33 was to ensure that the timer DOESN'T expire at the original set time, but was instead updated by the second arm failsafe call. So this shouldn't be here, but also the assert below should be failing if this failsafe has been expired

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,796 @@
+                             "Failure status returned from arm failsafe")
+
+        self.step(38)
+        # Step 38 - Skipped
```
keep this in for the cert testing - it is just on the CI that we're going to avoid waiting.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,796 @@
+        logger.info("Step 38 skipped: Saving Tstart (current wall time clock) is bypassed due to failsafe expiration workaround.")
+
+        self.step(39)
+        trusted_root_list_original_updated = await self.read_single_attribute_check_success(
```
this isn't from the test plan...

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,796 @@
+        logger.info(
+            f'Step #39 - The updated size of the num_trusted_roots_original list: {trusted_root_list_original_size_updated}')
+
+        # Flow generates a new TrustedRootCertificate - Request CSR (Certificate Signing Request) and update NOC (Node Operational Certificate)
```
can re-use the previously generated one.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,796 @@
+                             "Step #31 - Unexpected number of entries in the TrustedRootCertificates table after update")
+
+        self.step(41)
+        # Step 41 - Skipped
```
keep this in for cert testing, skip for CI.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,796 @@
+        # which immediately expires the failsafe timer (sets it to 0), bypassing the natural wait and speeding up the test execution.
+
+        # This function bypasses the wait for the FailsafeTimer to expire for TH1 as originally defined in the test plan.
+        resp = await self.expire_failsafe_timer(dev_ctrl=self.default_controller, node_id=self.dut_node_id)
```
gate on CI

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,796 @@
+        cmd = Clusters.OperationalCredentials.Commands.RemoveFabric(fabricIndex=fabric_idx)
+        await self.send_single_cmd(dev_ctrl=TH2, node_id=newNodeId+1, cmd=cmd)
+
+        # The expected number of root certificates should be 1 after removing the fabric
```
```suggestion
        # The expected number of root certificates should be numTrustedRootsOriginal after removing the fabric
```

juandediosg committed:
Code review done, changes and refactoring implemented

juandediosg committed:
Improving long waits for cert test

juandediosg committed:
Add PICS argument for CI

juandediosg committed:
Refactor code: improved flow handling and optimized wait times

juandediosg committed:
Refactor function call to is_pics_sdk_ci_only and update/improve waits

juandediosg committed:
Fixed: Lint Code Base

juandediosg committed:
Code cleanup and restyle

juandediosg commented:
Final version. Below are the highlighted updates, changes, and refactoring:

**Note:** The script has passed CI job checkpoints, including Lint code base, Restyled, etc.

- Refactored and cleaned up.
- Updated code and logger comments. 
- For the CI flow, the logic was implemented with the `is_pics_sdk_ci_only` to correctly handle test depending on environment.
- Adjusted the sequence between `maxFailSafe` and `PIXIT.CGEN.FailsafeExpiryLengthSeconds`.
- In the  CERT tests, `failsafe_expiration_seconds = 20` was set for the waits, while for CI, the `expire_failsafe_timer `function was used with a small re-arm of the timer (1 second) to correctly verify the timer expiration.
- Reused previously generated certificates.
- Steps 3, 4, and 5 were wrapped into one function with boolean control, while steps 8 and 9 were grouped into another function with boolean control.
- For step 22, the code remained unchanged, and it was suggested not to modify the `SetSkipCommissioningComplete(True)` logic, as it resulted in an error at step 29.
- Implemented `FindOrEstablishPASESession` function from ChipDeviceCtrl.py to avoid duplication.
- Mobly asserts were updated to improve error handling.
- Updated logic on sequence of steps 29-36 to align with the new optimizations.

jtrejoespinoza-grid commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,873 @@
+        resp = await TH2.FindOrEstablishPASESession(setupCode=setup_qr_code, nodeid=newNodeId)
+        logger.info('Step #22 - TH2 successfully establish PASE session completed')
+
+        try:
```
Maybe is a good idea to avoid using a generic Exception and use and more case specific exception, for example the previous comment is mentioned a timeout error, is good idea to catch that in specific to continue the steps from the test.

jtrejoespinoza-grid commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,873 @@
+
+        self.step(27)
+        # TH2 TrustedRootCertificate response with FAILSAFE_REQUIRED
+        try:
```
Similar here, try to avoid using a generic Exception and use the specific exception.

cecille replied:
There's also mobly support for expecting exceptions. 
with asserts.assert_raises(....):

https://github.com/google/mobly/blob/master/mobly/asserts.py#L263

I don't REALLY care if you use that vs. try / except, but if you use try / except, use the specific exception, AND you need to fail the test if the expected exception doesn't happen (ie, you need a fail line after the last command)

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,873 @@
+    cluster_opcreds = Clusters.OperationalCredentials
+    cluster_cgen = Clusters.GeneralCommissioning
+
+    async def expire_failsafe_timer(self, dev_ctrl, node_id, expiration_time_seconds):
```
naming - normally setting the failsafe to 0 is "expiring" or "disarming" the failsafe timer. Any non-zero value is normally "arming" the failsafe timer for that time. a failsafe of 0 is treated as an immediate disarm.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,873 @@
+            response: The response from the command sent to the device.
+        '''
+        # Buffer for latency
+        buffer_latency = .5
```
If you use 0 to disarm, you shouldn't need a buffer here.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,873 @@
+        if c_time is None:
+            c_time = time.time()
+
+        formatted_time = (
```
I'm guessing this is for logging, but does having the UTC time help here, or would it make more sense to have the number of elapsed ms (or us or whatever) since t_start?

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,873 @@
+        logger.info(f'Step #6: The updated size of the num_trusted_roots_original list: {trusted_root_list_original_size_updated}')
+
+        self.step(7)
+        if self.is_pics_sdk_ci_only:
```
I think this might actually be fine for ALL devices - you're still seeing the failsafe expire on its timer vs. forcing it to 0 explicitly. ie, it might make sense to just adjust the test to set this to 1s before waiting.

cecille replied:
yes, the test is just trying to ensure that the dut times out correctly vs. being forced to expire. We don't have to wait out the original time even in cert - you can set to this 1s for both CI and for real devices and the test results are still valid because the expiration happens as a result of a timer, even if the timer is small. Just add a test step to the test plan, and it cuts down the wait time for this test during both CI and certification.

juandediosg replied:
Fixed: To manage timeout behavior effectively, CI and Cert tests handled a bit differently: CI uses the PIXIT.CGEN.FailsafeExpiryLengthSeconds value for dynamic timeout adjustment based on the environment, while Cert uses the FAILSAFE_EXPIRATION_SECONDS constant for flexibility in setting longer timeouts if needed.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,873 @@
+    async def test_TC_CGEN_2_2(self):
+
+        # PIXIT.CGEN.FailsafeExpiryLengthSeconds:
+        # Timeout used in test steps to verify failsafe. Must be less than DUT MaxCumulativeFailsafeSeconds
```
Can you hook this up to a user parameter? https://project-chip.github.io/connectedhomeip-doc/testing/python.html#pics-and-pixits. that will let testers adjust this if required.

cecille replied:
Can you please add a default here? And a check that it's less than MaxCumulativeFailsafeSeconds

juandediosg replied:
Fixed: Added FAILSAFE_EXPIRATION_SECONDS constant value, for flexibility to adjust the timeout with longer waits if needed it on cert testing

juandediosg replied:
To manage timeout behavior effectively, CI and Cert tests handled a bit differently: CI uses the `PIXIT.CGEN.FailsafeExpiryLengthSeconds` value for dynamic timeout adjustment based on the environment, while Cert uses the `FAILSAFE_EXPIRATION_SECONDS` constant for flexibility in setting longer timeouts if needed.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,796 @@
+        logger.info('Step #15 - TH2 successfully establish PASE session completed')
+
+        self.step(16)
+        logger.info('Step #16 - TH2 Generating a new CSR to update the root certificate...')
```
I meant the process of generating a new cert. But actually, as long as you're always adding a cert that's not the one on the device, you should be able to re-use the first certificate instead of doing this dance every time.

cecille replied:
You should be able to re-use the same root certificate you created earlier to do this without sending commands to the device again.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,873 @@
+        logger.info(f'Step #20 - TH1 Original size of the trusted_root list: {trusted_roots_list_size}')
+
+        self.step(21)
+        # Commissioning stage numbers - we should find a better way to match these to the C++ code
```
It looks like the folks doing some of the new features plumbed through the functions required to do this using the commissioning parameters. Can you take a look at CGEN-2.5 (I James used it a few places). https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/TC_CGEN_2_5.py#L88. That will make this much neater, and I think it means you don't need to use the test commissioner.

juandediosg committed:
Fixed: Update try/except with specific exceptions and fail the test if the expected exception is not raised

juandediosg committed:
Merge branch 'project-chip:master' into TC_CGEN_2_2_PythonTest

juandediosg committed:
Fixed: Rename fn  to  for clarity based on feedback

juandediosg committed:
Fixed: Removed UTC time function, logging elapsed time in ms in steps 41 and 43.

juandediosg committed:
Merge branch 'project-chip:master' into TC_CGEN_2_2_PythonTest

jtrejoespinoza-grid reviewed (commented):
Changes requested updated.

jtrejoespinoza-grid reviewed (approved):
Mentioned issues addressed.

juandediosg committed:
Fixed: Hooked failsafe_expiration_seconds to user parameter PIXIT.CGEN.FailsafeExpiryLengthSeconds for CI or Cert flow flexibility.

juandediosg committed:
Fixed: Lint Code base and Restyled.

juandediosg committed:
Fixed: Argument PIXIT.CGEN.FailsafeExpiryLengthSeconds updated to 1

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,856 @@
+        # Verify that failsafe_timeout_less_than_max is less than max_fail_safe
+        asserts.assert_less(failsafe_timeout_less_than_max, maxFailsafe)
+
+        if self.is_pics_sdk_ci_only:
```
I'm not sure I see the difference here between the CI and the cert test.

juandediosg replied:
To manage timeout behavior more effectively, we approach CI and Cert tests differently. Both are currently set to 1 second, but they handle timeout a bit differently:

- **CI:** `PIXIT.CGEN.FailsafeExpiryLengthSeconds` PIXIT value, to adjust the timeout duration based on the CI and specs req.

- **[FIXED] Cert:** `FAILSAFE_EXPIRATION_SECONDS` constant value, for flexibility to adjust the timeout with longer waits if needed it.

The goal is to maintain flexibility to adjust the timeouts as needed, depending on the type of testing.

cecille replied:
The original comment still stands - in both of these cases, you're doing 
await asyncio.sleep(failsafe_timeout_less_than_max + .5)

juandediosg committed:
Added constant variable FAILSAFE_EXPIRATION_SECONDS for cert tests to handle timeout more effectively

juandediosg reviewed (commented):
New changes have been made and are ready for review.

juandediosg committed:
Add step 2a to configure failsafe timer for CI and Cert environments, and create constant RUN_TEST_ to define the run test type.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,871 @@
+            failsafe_expiration_seconds = self.matter_test_config.global_test_params['PIXIT.CGEN.FailsafeExpiryLengthSeconds']
+        else:
+            run_type = self.RUN_TEST_CERT
+            failsafe_expiration_seconds = self.CERT_FAILSAFE_EXPIRATION_TIME_SECONDS
```
This wasn't quite what I meant. This failsafe needs to be adjustable for cert. In fact, it's MORE important that it's adjustable for certification because this test will be run against devices that have variable network technologies and conditions. There should be a reasonable default.

So for BOTH CI and certification, you can use
default_failsafe = 20
failsafe_expiration_seconds = self.user_params.get(“PIXIT.CGEN.FailsafeExpiryLengthSeconds”, default_failsafe)

The point was that step 8 is attempting to ensure that the failsafe times out on its own (ie, that the DUT is setting the timer correctly). The time THERE doesn't matter. So instead of step 7 reading 
"TH1 waits for PIXIT.CGEN.FailsafeExpiryLengthSeconds to ensure the failsafe timer has expired"

Change that to 

"TH1 sends an ArmFailSafe command to the DUT with ExpiryLengthSeconds field set to 1 and the Breadcrumb value as 2. TH1 then waits 1.5 seconds to ensure the failsafe timer has expired"

juandediosg replied:
Fixed: I have updated the failsafe timer logic for both CI and certification environments: for both CI and certification, the default expiration time is 20 seconds, but it is adjustable through user parameters.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,871 @@
+            run_type = self.RUN_TEST_CI
+            failsafe_expiration_seconds = self.matter_test_config.global_test_params['PIXIT.CGEN.FailsafeExpiryLengthSeconds']
+        else:
+            run_type = self.RUN_TEST_CERT
```
This gets set a bunch of times during the test, but it won't change past this point. (ie, pics are fixed for the duration of the test)

juandediosg replied:
Fixed

juandediosg committed:
Fixed: Adjust failsafe timer for CI and Cert, with configurable expiration and default value of 20 seconds

juandediosg reviewed (commented):
I made the updates based on code review

juandediosg committed:
Fixed: Step 7 updated.

juandediosg committed:
Fixed: Step 2a updated in order to match with test plan.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,850 @@
+            f'Step #7: {run_type} - Forcing failsafe expiration by sending an ArmFailSafe command to the DUT with ExpiryLengthSeconds set to {failsafe_expiration_seconds} seconds.'
+        )
+
+        resp = await self.set_failsafe_timer(
```
Please just make this 1s for cert and CI and adjust the test plan.

juandediosg replied:
Fixed:  Updated to 1s for cert and CI

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,850 @@
+        # TODO: https://github.com/project-chip/connectedhomeip/issues/36629
+        kFindOperationalForCommissioningComplete = 30
+        logger.info(
+            f'Step #21 - TH2 Commissioning stage SetTestCommissionerPrematureCompleteAfter enum: {kFindOperationalForCommissioningComplete}')
```
Can you please verify that this does what you think it does? I believe SetPrematureCompleteAfter DOES send a commissioning complete command

juandediosg replied:
After reviewing the implementation, I can confirm that the `SetTestCommissionerPrematureCompleteAfter` function doesn’t send the `CommissioningComplete` command. Instead, configures the state for premature completion, which is exactly what this test scenario needs. 
This approach ensures the `CommissioningComplete` command isn’t sent and keeps the test behavior as expected.

cecille replied:
thanks.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,850 @@
+                node_id=newNodeId,
+                cmd=cmd)
+        except ChipStackError as e:
+            asserts.assert_in('Timeout',
```
Is this a test plan error? Test plan says it should respond with OK. Is this caused by the pase session getting evicted while you're using it?

juandediosg replied:
Yes, the issue comes from an incorrect expectation in the test plan for Step # 22, as it is expecting an "OK" response. However, the commissioning process is intentionally incomplete at Step # 21, as the `CommissioningComplete` command is not sent.

As a result, the PASE session is invalidated because the DUT is not fully commissioned. This leads to the `timeout error` when attempting to send further commands, as the session is no longer valid.

This is the reason I implemented the `except` block to handle the timeout error, which is the expected behavior due to the incomplete commissioning process.

I will update the Step # 22 to ensure that the implementation aligns with the expected behavior in the test plan (Current  updates on Test Plan PR [TC_CGEN_2_2 Test Plan Updated - Test Procedure and Table (new format) #4956](https://github.com/CHIP-Specifications/chip-test-plans/pull/4956)) as Open.

cecille replied:
step 22 isn't updated in that PR.

cecille replied:
I'm worried this is too SDK specific - this is testing all devices against what the SDK does. Really, the verification that the failsafe was successfully disarmed happens below, so can you just change this to be except: pass with an explanation?

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,850 @@
+                             "The trusted_roots list size should match the original trusted_roots list size.")
+
+        self.step(26)
+        basic_commissioning_info = await self.read_single_attribute_check_success(cluster=Clusters.GeneralCommissioning, attribute=Clusters.GeneralCommissioning.Attributes.BasicCommissioningInfo)
```
What's basic_commissioning_info used for?

juandediosg replied:
The `basic_commissioning_info` variable is used only for logging the `BasicCommissioningInfo` attribute during the test and does not affect any logic or assertions. 
Please let me know this log entry is unnecessary, and I can remove it.

cecille replied:
yeah, please remove.

juandediosg replied:
Fixed: Logger removed.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,850 @@
+        try:
+            logger.info("Step #27 - TH2 Generating a new CSR to update the root certificate...")
+            # Flow generates a new TrustedRootCertificate - Request CSR (Certificate Signing Request) and update NOC (Node Operational Certificate)
+            cmd = self.cluster_opcreds.Commands.CSRRequest(CSRNonce=random.randbytes(32), isForUpdateNOC=False)
```
same here - you can re-use the first dummy RCAC to avoid sending additional commands to the device.

juandediosg replied:
Fixed: Used the first RCAC (`new_root_cert`)

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,850 @@
+        logger.info(f'Step #37: ArmFailSafeResponse with ErrorCode as OK({resp.errorCode})')
+
+        self.step(38)
+        if self.is_pics_sdk_ci_only:
```
I think after this point, none of the steps apply for the CI, so you might as well just bail out here. There's a function to mark the remaining steps as skipped.

juandediosg replied:
Currently, the test is working as expected. I have skipped certain steps in the CI to avoid delays, particularly during failsafe expiration waits like steps # 41 and # 43, which have been bypassed in CI to optimize execution time (based on previous feedback), while local tests run all steps.

However, if you would like me to refactor/update the code to skip steps from # 38 onward in the CI, please let me know.

cecille replied:
yes, please just skip the remaining steps for the CI - right now, the test works, but it's hard to understand what's going on.

juandediosg replied:
Fixed: Bypass steps from #38 onward ONLY in CI.

juandediosg committed:
Fixes and updates from feedback

juandediosg reviewed (commented):
Please find below a highlight of the fixes/updates based on feedback:

- **Step # 7 - Update set_failsafe_timer:** Updated to 1s for cert and CI. 
Test Plan need to be updated on current PR [TC_CGEN_2_2 Test Plan Updated - Test Procedure and Table (new format) #4956)](https://github.com/CHIP-Specifications/chip-test-plans/pull/4956).

- **Step # 21 - Establishing a CASE connection, but DOES NOT send the CommissioningComplete command:** Confirm that the `SetTestCommissionerPrematureCompleteAfter` function doesn’t send the `CommissioningComplete` command and this approach ensures the `CommissioningComplete` command isn’t sent and keeps the test behavior as expected.

- **Step # 22 - Timeout:** The `Timeout` is expected, instead of the "OK" as showed in test Plan. Since the commissioning process is intentionally incomplete at Step # 21 (no `CommissioningComplete` command is sent), the implementation reflect the expected behavior.
Test Plan need to be updated on current PR [TC_CGEN_2_2 Test Plan Updated - Test Procedure and Table (new format) #4956)](https://github.com/CHIP-Specifications/chip-test-plans/pull/4956).

- **Step # 26 - basic_commissioning_info:** The `basic_commissioning_info` variable is used only for logging the `BasicCommissioningInfo` attribute during the test and doesn't affect any logic or assertions. If this log entry is unnecessary, I can remove it. 

- **Step # 27 - Reuse RCAC:** Updated to reuse the first RCAC (new_root_cert).

- **Step # 38 - Skip steps from # 38 onward in the CI:** In the CI environment, I have skipped certain steps (such as # 41 and # 43) to avoid unnecessary delays. These steps were bypassed in CI to optimize execution time, based on previous feedback, while local tests still execute all steps to maintain the full commissioning process. If you would prefer to refactor the code to skip steps from # 38 onward in CI, please let me know.

- **Buffer Removed:** The buffer of .5s was removed.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,838 @@
+        resp = await self.set_failsafe_timer(
+            dev_ctrl=self.default_controller,
+            node_id=self.dut_node_id,
+            expiration_time_seconds=failsafe_expiration_seconds)
```
this isn't 1s though - this is whatever the pixit value is set to.

juandediosg replied:
I have set the `expiration_time_seconds` to 1s for step # 7 as requested. 

Additionally, I made a minor adjustment on the PIXIT configuration because the `failsafe_expiration_seconds` that comes from PIXIT is still utilized in other steps.

Would you prefer that I completely remove PIXIT and set `failsafe_expiration_seconds` to 1 instead? Just as friendly comment, I originally added this PIXIT (`PIXIT.CGEN.FailsafeExpiryLengthSeconds`) based on previous feedback to allow this value to be adjustable as part of the improvements.

juandediosg committed:
Fixed: Step #26 logger removed.

juandediosg committed:
Fixed: Bypass steps from #38 onward ONLY in CI.

juandediosg committed:
Fixed: Set expiration_time_seconds=1 in Step #7, and minor adjustments to PIXIT.CGEN.FailsafeExpiryLengthSeconds.

juandediosg reviewed (commented):
Please find below a highlight of the fixes/updates based on feedback:

**Step # 7**
Fixed: Set `expiration_time_seconds=1`.
Also, a minor update on PIXIT to use 1s when no value is provided and still keeps adjustable. And still ensured failsafe_expiration_seconds remains configurable

**Step # 26**
Fixed: logger `basic_commissioning_info` has been removed

**Step # 38**
Fixed: Bypass steps from # 38 onward ONLY in CI.

Also new push to the Test Plan PR [TC_CGEN_2_2 Test Plan Updated - Test Procedure and Table (new format) #4956
](https://github.com/CHIP-Specifications/chip-test-plans/pull/4956) with the below updates:
- Updated **Step # 7** to set ExpiryLengthSeconds to 1.
- Updated **Step # 22** to expect a timeout error (`CHIP Error 0x00000032`) instead of `OK`, due to incomplete commissioning.

juandediosg committed:
Re-use root certificate from step 5 in steps 16, 27, 30, and 39

juandediosg committed:
Fixed: Step 33 Simplified logic, removed CI/Cert flows

juandediosg reviewed (commented):
Hi @cecille ,

Please find below the highlighted fixes and updates made based on your comments:

**Step 16 Fixed:** Reused the root certificate from step 5 in steps 16, 27, 30, and 39.

**Step 22 - Updated Test Plan PR:** Updated Step 22 to expect a timeout error (`CHIP Error 0x00000032`) instead of `OK`, due to incomplete commissioning.
Commit [238abb8c4ad7f4234c93632550b10879be2e36c2](https://github.com/CHIP-Specifications/chip-test-plans/pull/4956/commits/238abb8c4ad7f4234c93632550b10879be2e36c2).

**Step 33 - Fixed:** Simplified logic by removing CI/Cert flow distinction.

juandediosg committed:
Fixed: PIXIT.CGEN.FailsafeExpiryLengthSeconds hard coded to 1.

juandediosg reviewed (commented):
I have updated the value of `pixit_failsafe_expiry_length_seconds` to 1 to match the expected test condition.

juandediosg committed:
Updated: Step 22 to handle expected behavior.

juandediosg reviewed (commented):
Updated Step 22 to handle expected behavior with `except: pass` as suggested, removing specific exception check

juandediosg committed:
Fixed: code-lints, removed unused import chip.exceptions.ChipStackError.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,813 @@
+                dev_ctrl=TH2,
+                node_id=newNodeId,
+                cmd=cmd)
+        except Exception as e:
```
Can you tighten this up to a ChipStackException? Exception is pretty broad.

juandediosg replied:
Fixed: Implemented `ChipStackException` instead of `Exception`.

cecille commented on src/python_testing/TC_CGEN_2_2.py:
```diff
@@ -0,0 +1,813 @@
+            pass
+        else:
+            # If no exception is raised, something is wrong — expected command to fail.
+            asserts.assert_true(False, 'Expected ArmFailSafe to fail because the device is not fully commissioned.')
```
I don't think this is necessary - if the device sends the response before it closes down the connection, I think that's also fine.

juandediosg replied:
Fixed: Removed unnecessary assert.

juandediosg committed:
Fixed: Use ChipStackException and remove unnecessary assert.

juandediosg reviewed (commented):
Please find below the updates/fixes based on comments/feedback in step 22:
- Implemented `ChipStackException` instead of `Exception`.
- Removed unnecessary `assert`.
